File: backend/app/schemas/identity_link.py
"""Schemas for identity links (#985) — see app.services.identity_links."""

from pydantic import BaseModel, Field


class ProposeIdentityLinkRequest(BaseModel):
    target_workspace_id: str
    target_member_id: str


class DecideIdentityLinkRequest(BaseModel):
    # Matches IdentityLink.decision_reason (String(500)) so an over-length
    # value is rejected as a 422 here instead of failing the write.
    reason: str | None = Field(default=None, max_length=500)


class RejectIdentityLinkRequest(DecideIdentityLinkRequest):
    # Also blocks the proposer from proposing further links into the
    # rejecting side's workspace.
    block_proposer: bool = False


class IdentityLinkEndpointOut(BaseModel):
    workspace_id: str
    workspace_name: str
    member_id: str
    display_name: str | None


class IdentityLinkOut(BaseModel):
    id: str
    status: str
    verification_basis: str
    self: IdentityLinkEndpointOut
    # None when the counterpart is degraded to a protected placeholder because
    # the viewer currently lacks read access to that workspace (#985) — the
    # link itself is untouched, only this rendering hides who it points to.
    counterpart: IdentityLinkEndpointOut | None
    counterpart_protected: bool
    proposed_at: str
    expires_at: str | None
    verified_at: str | None
    decided_at: str | None
    decision_reason: str | None


class IdentityLinkListOut(BaseModel):
    links: list[IdentityLinkOut]


class ProposeIdentityLinkClaimRequest(BaseModel):
    # An accepted friend's username — never a workspace or member id, so the
    # proposer never needs (or gets) any visibility into the target's
    # workspace (#1014's opaque claim/invitation flow).
    target_username: str = Field(min_length=1, max_length=100)
    # Free text about the proposer's *own* member, shown to the target to
    # help them decide — never anything read from the target's workspace.
    note: str | None = Field(default=None, max_length=500)


class CompleteIdentityLinkClaimRequest(BaseModel):
    workspace_id: str
    member_id: str


class DecideIdentityLinkClaimRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=500)


class IdentityLinkClaimOut(BaseModel):
    id: str
    status: str
    source_workspace_id: str
    source_workspace_name: str
    source_member_id: str
    source_display_name: str | None
    proposer_username: str | None
    target_username: str
    note: str | None
    created_at: str
    expires_at: str | None
    decided_at: str | None
    decision_reason: str | None
    resulting_identity_link_id: str | None


class IdentityLinkClaimListOut(BaseModel):
    claims: list[IdentityLinkClaimOut]

File: frontend/src/types/identityLink.ts
export type IdentityLinkStatus =
  "proposed" | "verified" | "rejected" | "expired" | "revoked";

export type IdentityLinkVerificationBasis =
  "mutual_consent" | "same_owner" | "legacy_dual_write_access";

export interface IdentityLinkEndpoint {
  workspace_id: string;
  workspace_name: string;
  member_id: string;
  display_name: string | null;
}

/** A verified or in-progress cross-workspace identity link, rendered from
 *  the viewer's own member's side (see `self`/`counterpart`). */
export interface IdentityLink {
  id: string;
  status: IdentityLinkStatus;
  verification_basis: IdentityLinkVerificationBasis;
  self: IdentityLinkEndpoint;
  /** Null when the viewer currently lacks read access to the counterpart —
   *  render `counterpart_protected` as a placeholder instead. */
  counterpart: IdentityLinkEndpoint | null;
  counterpart_protected: boolean;
  proposed_at: string;
  expires_at: string | null;
  verified_at: string | null;
  decided_at: string | null;
  decision_reason: string | null;
}

export type IdentityLinkClaimStatus =
  "pending" | "completed" | "declined" | "cancelled" | "expired";

/** The opaque claim/invitation half of the proposal flow: the proposer names
 *  only their own member and a friend, who later picks their own member. */
export interface IdentityLinkClaim {
  id: string;
  status: IdentityLinkClaimStatus;
  source_workspace_id: string;
  source_workspace_name: string;
  source_member_id: string;
  source_display_name: string | null;
  proposer_username: string | null;
  target_username: string;
  note: string | null;
  created_at: string;
  expires_at: string | null;
  decided_at: string | null;
  decision_reason: string | null;
  resulting_identity_link_id: string | null;
}
